### DISCUSSION TITLE
More semantics, in a structured and strategic way!

### DISCUSSION THREAD

**Discussion by pdehaye** [hestiaAI/Argonodes#58] [General]
This diagram is a visual description of what we are doing right now ([source](https://q.uiver.app/?q=WzAsMjQsWzIsMCwiXFx0ZXh0e2FjY2Vzc29yIGV4dHJhY3Rpb259ICJdLFs1LDEsIlxcdGV4dHtTUUwgcXVlcnl9Il0sWzcsMSwiXFx0ZXh0e3Zpen0iXSxbMCwxLCJcXHRleHR7ZHJvcH0iXSxbNCwxLCJcXGJ1bGxldCJdLFsxLDEsIlxcYnVsbGV0Il0sWzMsMiwiXFx0ZXh0e2RiIGltcG9ydCBmcm9tIGNzdn0iXSxbMTEsMSwiXFx0ZXh0e2NvbnNlbnR9Il0sWzEzLDEsIlxcdGV4dHt0cmFuc2Zlcn0iXSxbOSwxLCJcXHRleHR7ZmlsdGVyfSJdLFsxNSwxLCJcXHRleHRiZntub3JtYWwgbGFuZH0iXSxbMiw1LCJcXHRleHR7fSJdLFszLDYsIlxcdGV4dHtkYiBpbXBvcnQgZnJvbSBjc3Z9Il0sWzQsNSwiXFxidWxsZXQiXSxbNSw1LCJcXHRleHR7U1FMIHF1ZXJ5fSJdLFs3LDUsIlxcdGV4dHt2aXp9Il0sWzksNSwiXFx0ZXh0e2ZpbHRlcn0iXSxbMTEsNSwiXFx0ZXh0e2NvbnNlbnR9Il0sWzEzLDUsIlxcdGV4dHt0cmFuc2Zlcn0iXSxbNywyXSxbMTUsNSwiXFx0ZXh0YmZ7TGlua2VkIERhdGEgbGFuZH0iXSxbMyw0XSxbNiwxLCJcXGJ1bGxldCJdLFszLDAsIlxcdGV4dHtkYiBpbXBvcnR9Il0sWzEsMiwiLmNzdiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19LFsyNDAsNjAsNjAsMV1dLFs0LDEsIi5zcWwiLDAseyJjb2xvdXIiOlszMCw2MCw2MF19LFszMCw2MCw2MCwxXV0sWzMsNV0sWzUsMCwiLmpzb24iLDAseyJjb2xvdXIiOlszMDAsNjAsNjBdfSxbMzAwLDYwLDYwLDFdXSxbNiw0LCIiLDAseyJjb2xvdXIiOlszMCw2MCw2MF19XSxbNyw4LCIiLDAseyJjb2xvdXIiOlsyNDAsNjAsNjBdfV0sWzIsOSwiIiwwLHsiY29sb3VyIjpbMjQwLDYwLDYwXX1dLFs5LDcsIiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19XSxbOCw2LCIiLDAseyJjdXJ2ZSI6LTUsImNvbG91ciI6WzI0MCw2MCw2MF19XSxbMTIsMTMsIiIsMCx7ImNvbG91ciI6WzMwLDYwLDYwXX1dLFsxMywxNCwiIiwwLHsiY29sb3VyIjpbMzAsNjAsNjBdfV0sWzE0LDE1LCIuY3N2TEQiLDAseyJjb2xvdXIiOlsyNDAsNjAsNjBdfSxbMjQwLDYwLDYwLDFdXSxbMTYsMTcsIiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19XSxbMTcsMTgsIiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19XSxbMTUsMTYsIiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19XSxbMTgsMTIsIiIsMCx7ImN1cnZlIjotNSwiY29sb3VyIjpbMjQwLDYwLDYwXX1dLFsxMCwyMCwiTEQoLikiLDAseyJzaG9ydGVuIjp7InNvdXJjZSI6MTAsInRhcmdldCI6MTB9LCJsZXZlbCI6Mn1dLFswLDIzLCI/Il0sWzUsNiwiLmNzdiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19LFsyNDAsNjAsNjAsMV1dLFsyMyw0LCIiLDAseyJjb2xvdXIiOlszMCw2MCw2MF19XSxbMjQsMzUsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMzAsMzgsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMzEsMzYsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMjksMzcsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMzIsMzksIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XV0=)) [work in progress](https://q.uiver.app/?q=WzAsMjIsWzIsMiwiXFx0ZXh0e2FjY2Vzc29yIGV4dHJhY3Rpb259ICJdLFs3LDEsIlxcdGV4dHtTUUwgcXVlcnl9Il0sWzEwLDYsIlxcdGV4dHt2aXp9Il0sWzAsMywiXFx0ZXh0e2RyYWcnbidkcm9wfSJdLFsxLDMsIlxcYnVsbGV0Il0sWzQsMywiXFx0ZXh0e2RiIGltcG9ydCBmcm9tIGNzdn0iXSxbNSw4LCJcXHRleHR7Y29uc2VudH0iXSxbNCw0LCJcXHRleHR7dHJhbnNmZXJ9Il0sWzksMiwiXFx0ZXh0e2ZpbHRlcn0iXSxbMTYsMywiXFx0ZXh0YmZ7bm9ybWFsIGxhbmR9Il0sWzMsNywiXFx0ZXh0e30iXSxbNCwxMCwiXFx0ZXh0e2RiIGltcG9ydCBmcm9tIGNzdn0iXSxbNyw4LCJcXHRleHR7U1FMIHF1ZXJ5fSJdLFsxMCwxMywiXFx0ZXh0e3Zpen0iXSxbOSw5LCJcXHRleHR7ZmlsdGVyfSJdLFs1LDE1LCJcXHRleHR7Y29uc2VudH0iXSxbNCwxMSwiXFx0ZXh0e3RyYW5zZmVyfSJdLFs4LDRdLFsxNiwxMCwiXFx0ZXh0YmZ7TGlua2VkIERhdGEgbGFuZH0iXSxbNCw2XSxbMywzLCJcXGJ1bGxldCJdLFsyLDAsIlxcdGV4dHtjb25maWd9Il0sWzEsMiwiLmNzdiIsMCx7ImN1cnZlIjozLCJjb2xvdXIiOlsyNDAsNjAsNjBdfSxbMjQwLDYwLDYwLDFdXSxbMyw0XSxbNCwwLCIuanNvbiIsMCx7ImN1cnZlIjotMiwiY29sb3VyIjpbMzAwLDYwLDYwXX0sWzMwMCw2MCw2MCwxXV0sWzYsNywiIiwwLHsiY29sb3VyIjpbMjQwLDYwLDYwXX1dLFs3LDUsIiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19XSxbMTIsMTMsIi5jc3ZMRCIsMCx7ImN1cnZlIjozLCJjb2xvdXIiOlsyNDAsNjAsNjBdfSxbMjQwLDYwLDYwLDFdXSxbMTUsMTYsIiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19XSxbMTYsMTEsIiIsMCx7ImNvbG91ciI6WzI0MCw2MCw2MF19XSxbOSwxOCwiTEQoLikiLDAseyJzaG9ydGVuIjp7InNvdXJjZSI6MTAsInRhcmdldCI6MTB9LCJsZXZlbCI6Mn1dLFs0LDIwLCIuY3N2IiwwLHsiY3VydmUiOjUsImNvbG91ciI6WzI0MCw2MCw2MF19LFsyNDAsNjAsNjAsMV1dLFsyMCw1LCIiLDAseyJjb2xvdXIiOlsyNDAsNjAsNjBdfV0sWzAsMjAsIi5jc3YiLDAseyJjdXJ2ZSI6LTIsImNvbG91ciI6WzI0MCw2MCw2MF19LFsyNDAsNjAsNjAsMV1dLFsxMSwxMiwiIiwwLHsiY29sb3VyIjpbMzAsNjAsNjBdfV0sWzUsMSwiLnNxbCIsMCx7ImNvbG91ciI6WzMwLDYwLDYwXX0sWzMwLDYwLDYwLDFdXSxbMjEsMF0sWzIsNiwiIiwwLHsiY29sb3VyIjpbMjQwLDYwLDYwXX1dLFsxMywxNSwiIiwwLHsiY29sb3VyIjpbMjQwLDYwLDYwXX1dLFsxLDgsIiIsMCx7ImN1cnZlIjotMiwiY29sb3VyIjpbMjQwLDYwLDYwXX1dLFsxMiwxNCwiIiwwLHsiY3VydmUiOi0yLCJjb2xvdXIiOlsyNDAsNjAsNjBdfV0sWzgsMiwiIiwwLHsiY3VydmUiOi0yLCJjb2xvdXIiOlsyNDAsNjAsNjBdfV0sWzE0LDEzLCIiLDAseyJjdXJ2ZSI6LTIsImNvbG91ciI6WzI0MCw2MCw2MF19XSxbMjIsMjcsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMjUsMjgsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMjYsMjksIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMzcsMzgsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbMzksNDAsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XSxbNDEsNDIsIiIsMCx7InNob3J0ZW4iOnsic291cmNlIjoxMCwidGFyZ2V0IjoxMH19XV0=). 

![image](https://user-images.githubusercontent.com/3274335/173793490-b0f88840-8146-4ea3-8e04-03e9a4040af2.png)

Each "land" has been built as a dataflow diagram, following the "meringue idea" of Applied Category Theory. 

Have a look at [this source, ](https://arxiv.org/abs/1809.05923).
![image](https://user-images.githubusercontent.com/3274335/173788763-9c4474ee-b145-48e5-ac7e-d032b9ede0f0.png)

There are at least two things I have not explained this far:
* why the resource (data, raw eggs) are on the arrows in the horizontal planes/lands.
* why the arrows between lands are double arrows. 

The two explanations are kind of orthogonal to each other, so it is worth delaying a bit.

**Comment by pdehaye** [comment]
Several questions came in: What do the down arrows mean? and What is the bottom land?

**Reply by pdehaye** [reply]
have added/changed slightly some labels, that should clear up those questions

**Reply by pdehaye** [reply]
is this clearer now, @alexbfree ?

**Comment by alexbfree** [comment]
Also 'drop' and 'transfer' could be clearer... and why do CSVs get imported to a db but json don't?

**Reply by valentinoli** [reply]
> why do CSVs get imported to a db but json don't?

Yes, it should be
JSON -> accessors -> db import -> sql query

**Reply by pdehaye** [reply]
"drop" is "drag&drop", "transfer" is the transfer between bubbles. I know that requires work on both bubbles, but that is not what I want to highlight in this diagram (this is where maybe I am most different in my concerns than your architecture vision). 

Have better suggestions?

**Reply by pdehaye** [reply]
have also updated the "db import side", but it is akward when you don't want to overlook important elements: I want the loop at the bottom to be clear and unambiguous, and "db import" is not the same if the import types are not the same. 

maybe a way forward to make it less awkward is to see accessors as a way to go from JSON to CSV ahead of database import. is this what you guys are doing?

**Reply by valentinoli** [reply]
> maybe a way forward to make it less awkward is to see accessors as a way to go from JSON to CSV ahead of database import. is this what you guys are doing?

yes

**Reply by pdehaye** [reply]
so you go through accessors from json to csv, and you go from csv to sql. nice. ok, time to redraw (quite substantially) this thing.

**Reply by emmanuel-hestia** [reply]
Hmmm we do not have an explicit JSON -> CSV step before SQL, it is all loaded directly from JSON to SQL. But it is easy to retrieve as CSV from the SQL, of course

**Comment by Amustache** [comment]
Related to https://github.com/hestiaAI/hestialabs-experiences/issues/767
